ext/web/webtransport.js

```diff
@@ -0,0 +1,1125 @@
+  }
+
+  set sendGroup(value) {
+    webidl.assertBranded(this, WebTransportSendStreamPrototype);
+    value = webidl.converters.WebTransportSendGroup(
+      value,
+      "Failed to execute 'sendGroup' on 'WebTransportSendStream'",
+    );
+    this.#sendGroup = value;
+  }
+
+  get sendOrder() {
+    webidl.assertBranded(this, WebTransportSendStreamPrototype);
+    return this.#inner.sendOrder;
+  }
+
+  set sendOrder(sendOrder) {
+    webidl.assertBranded(this, WebTransportSendStreamPrototype);
+    sendOrder = webidl.converters["long long"](
+      sendOrder,
+      "Failed to execute 'sendOrder' on 'WebTransportSendStream'",
+    );
+    this.#inner.sendOrder = sendOrder;
+  }
+
+  getStats() {
+    webidl.assertBranded(this, WebTransportSendStreamPrototype);
+    return PromiseResolve({
+      bytesWritten: 0,
+      bytesSent: 0,
+      bytesAcknowledged: 0,
+    });
+  }
+
+  getWriter() {
+    webidl.assertBranded(this, WebTransportSendStreamPrototype);
+    return new WebTransportWriter(this);
+  }
+}
+webidl.configureInterface(WebTransportSendStream);
+const WebTransportSendStreamPrototype = WebTransportSendStream.prototype;
+
+class WebTransportReceiveStream extends ReadableStream {
+  [webidl.brand] = webidl.brand;
+  #inner;
+
+  constructor(brand, key, inner) {
+    if (key !== illegalConstructorKey) {
+      webidl.illegalConstructor();
+    }
+    super(brand);
+    this.#inner = inner;
+  }
+
+  getStats() {
+    webidl.assertBranded(this, WebTransportReceiveStreamPrototype);
+    return PromiseResolve({
+      bytesReceived: 0,
+      bytesRead: 0,
+    });
+  }
+}
+webidl.configureInterface(WebTransportReceiveStream);
+const WebTransportReceiveStreamPrototype = WebTransportReceiveStream.prototype;
+
+class WebTransportWriter extends WritableStreamDefaultWriter {
+  [webidl.brand] = webidl.brand;
+
+  // atomicWrite() {}
+}
+webidl.configureInterface(WebTransportWriter);
+
+class WebTransportDatagramDuplexStream {
+  [webidl.brand] = webidl.brand;
+  #promise;
+  #conn;
+  #sessionIdBuf;
+  #readable;
+  #readableController;
+  #writable;
+  #incomingMaxAge = Infinity;
+  #outgoingMaxAge = Infinity;
+  #incomingHighWaterMark = 1;
+  #outgoingHighWaterMark = 5;
+  #incomingDatagramsPullPromise = null;
+  #incomingDatagramsQueue = [];
+  #outgoingDatagramsQueue = [];
+  #sending = false;
+
+  constructor(key, promise) {
+    if (key !== illegalConstructorKey) {
+      webidl.illegalConstructor();
+    }
+
+    this.#promise = promise;
+    PromisePrototypeThen(promise, ({ conn, sessionIdBuf }) => {
+      this.#conn = conn;
+      this.#sessionIdBuf = sessionIdBuf;
+    });
+
+    this.#receiveDatagrams();
+  }
+
+  async #receiveDatagrams() {
+    const { conn, sessionIdBuf } = await this.#promise;
+    while (true) {
+      const queue = this.#incomingDatagramsQueue;
+      const duration = this.#incomingMaxAge ?? Infinity;
+
+      let datagram;
+      try {
+        datagram = await conn.readDatagram();
```

Comment:
Shouldn't this one be unrefed? Won't it cause leaks in tests?

Reply:
when the connection is closed, this raises. so it breaks the loop. test does not detect any leaks.